onmt/transforms/tokenize.py

```diff
@@ -208,59 +249,43 @@ def warm_up(self, vocabs=None):
                 )
             self.load_models = {"src": load_src_model, "tgt": load_tgt_model}
 
-    def _tokenize(self, tokens, side="src", is_train=False):
-        """Do sentencepiece subword tokenize."""
+    def segment_with_subword_sampling(self, nbest_size, string, side="src"):
+        print("sampling ...")
         sp_model = self.load_models[side]
-        sentence = " ".join(tokens).replace(DefaultTokens.SEP, "\n")
-        sent_list = sentence.split(DefaultTokens.EOS)
-        segmented = []
-        for sentence in sent_list:
-            nbest_size = (
-                self.tgt_subword_nbest if side == "tgt" else self.src_subword_nbest
-            )
-            if is_train is False or nbest_size in [0, 1]:
-                # derterministic subwording
-                segmented += sp_model.encode(sentence, out_type=str)
-            else:
-                # subword sampling when nbest_size > 1 or -1
-                # alpha should be 0.0 < alpha < 1.0
-                alpha = (
-                    self.tgt_subword_alpha if side == "tgt" else self.src_subword_alpha
-                )
-                segmented += sp_model.encode(
-                    sentence,
-                    out_type=str,
-                    enable_sampling=True,
-                    alpha=alpha,
-                    nbest_size=nbest_size,
-                )
-            segmented += [DefaultTokens.EOS]
+        alpha = self.tgt_subword_alpha if side == "tgt" else self.src_subword_alpha
+        print(self.src_subword_alpha)
+        tokens = sp_model.encode(
+            string,
+            out_type=str,
+            enable_sampling=True,
+            alpha=alpha,
+            nbest_size=nbest_size,
+        )
+        return tokens
 
-        return segmented[:-1]
+    def segment_determistic(self, string, side="src"):
+        print("not sampling ...")
+        sp_model = self.load_models[side]
+        tokens = sp_model.encode(string, out_type=str)
+        return tokens
+
+    def tokenize_string(self, string, side="src", is_train=False):
+        """Apply subword sampling or deterministic subwording"""
+        nbest_size = self.tgt_subword_nbest if side == "tgt" else self.src_subword_nbest
+        if is_train is False or nbest_size in [0, 1]:
+            # derterministic subwording
+            tokens = self.segment_determistic(string, side)
```

Comment:
Maybe we should keep "tokenize_xx" instead of "segment_xx", as segment could be confusing with "sentence/paragraph" segmentation rather than the segmentation of tokens here.